File: submissions/Chimoney-Python/chimoney/Payouts.py
from chimoney import BaseAPI


class Payouts(BaseAPI):
    """
    This Wraps the Payouts API of Chi Money
    """

    def airtime(self, airtimes, subaccount=None, turn_off_notification=None):
        """
        This function handles the airtime API.

        Args:
            airtimes(list, required): A list of dictionaries containing the airtime details.
            subaccount(str): The subaccount to use.
            turn_off_notificaiton(bool): None by default

        example:
            airtime = [
                {
                   countryToSend: "Nigeria",
                   phoneNumber: "+2348123456789",
                   valueInUSD: 123
                }
            ]

        Return:
            The JSON response from the Chi Money API.
        """

        if not isinstance(airtimes, list):
            raise ValueError("airtimes must be a list of dictionaries")

        payload = {"airtimes": airtimes}
        if subaccount:
            payload["subAccount"] = subaccount

        if turn_off_notification is not None:
            payload["turnOffNotification"] = turn_off_notification

        return self._handle_request("POST", "/v0.2/payouts/airtime", data=payload)

    def bank(self, banks, subaccount=None, turn_off_notification=None):
        """
        This function handles the bank API.

        Args:
            banks(list, required): A list of dictionaries containing the bank details.
            subaccount(str): The subaccount to use.
            turn_off_notificaiton(bool): None by default

        example:
            banks = [
                {
                    "countryToSend": "Nigeria",
                    "account_bank": "044",
                    "account_number": "0690000031",
                    "valueInUSD": 1,
                    "reference": "1234567890",
                    "fullname": "James John",
                    "branch_code": "GH190101"
                }
            ]

        Returns:
            The JSON response from the Chi Money API.
        """

        if not isinstance(banks, list):
            raise ValueError("banks must be a list of dictionaries")

        payload = {"banks": banks}
        if subaccount:
            payload["subAccount"] = subaccount

        if turn_off_notification is not None:
            payload["turnOffNotification"] = turn_off_notification

        return self._handle_request("POST", "/v0.2/payouts/bank", data=payload)

    def chimoney(self, chimoneys, subaccount=None, turn_off_notification=None):
        """
        This function handles the chimoney API.

        Args:
            chimoneys(list, required): A list of dictionaries containing the chimoney details.
            subaccount(str): The subaccount to use.
            turn_off_notification(bool): None by Default

        example:
            chimoneys = [
                {
                    "valueInUSD": 1,
                    "email": "test@example.com",
                    "phone": "+2341234567890",
                }
            ]

        Return:
        The response from the Chi Money API.
        """

        if not isinstance(chimoneys, list):
            raise ValueError("chimoney must be a list of dictionaries")

        payload = {"chimoneys": chimoneys}
        if subaccount:
            payload["subAccount"] = subaccount

        if turn_off_notification is not None:
            payload["turnOffNotification"] = turn_off_notification

        return self._handle_request("POST", "/v0.2/payouts/chimoney", data=payload)

    def mobile_money(self, momos, subaccount=None, turn_off_notification=None):
        """
        This function handles the mobile money API.

        Args:
            momos(list): A list of dictionaries containing the mobile money details.
            subaccount(str): The subaccount to be used.
            turn_off_notification(bool): None by Default

        example:
            momos = [
                {
                    "countryToSend": "Nigeria",
                    "phoneNumber": "+2348123456789",
                    "valueInUSD": 1,
                    "reference": "1234567890",
                    "momoCode": "MPS"
                }
            ]

        Return:
            The JSON response from the Chi Money API.
        """

        if not isinstance(momos, list):
            raise ValueError("mobile_money must be a list of dictionaries")

        payload = {"momos": momos}
        if subaccount:
            payload["subAccount"] = subaccount

        if turn_off_notification is not None:
            payload["turnOffNotification"] = turn_off_notification

        return self._handle_request("POST", "/v0.2/payouts/mobile-money", data=payload)

    def gift_card(self, gift_cards, subaccount=None, turn_off_notification=None):
        """
        This function handles the gift card API.

        Args:
            gift_cards(list): A list of dictionaries containing the gift card details.
            subaccount(str): The subaccount to use.
            turn_off_notification(bool): None by Default.

        example:
            gift_cards = [
                {
                    "email": "test@example.com",
                    "valueInUSD": 1,
                    "redeemData": {
                        "productId": 5,
                        "countryCode": "NG"
                        "valueInLocalCurrency": 1000
                    }
                }
            ]

        Return:
            The JSON response from the Chi Money API.
        """

        if not isinstance(gift_cards, list):
            raise ValueError("gift_cards must be a list of dictionaries")

        payload = {"giftcards": gift_cards}
        if subaccount:
            payload["subAccount"] = subaccount

        if turn_off_notification is not None:
            payload["turnOffNotification"] = turn_off_notification

        return self._handle_request("POST", "/v0.2/payouts/gift-card", data=payload)

    def status(self, chi_ref, subaccount=None):
        """
        This function handles the status API.

        Args:
            chi_ref(str): The Chi Money reference.
            subaccount(str): The subaccount to use.

        Return:
            dict: The response from the Chimoney API.
        """

        if not chi_ref:
            raise ValueError("chi_ref is required")

        payload = {"chiRef": chi_ref}
        if subaccount:
            payload["subAccount"] = subaccount

        return self._handle_request("POST", "/v0.2/payouts/status", data=payload)

    def initiate_chimoney(
        self,
        chimoneys,
        crypto_payments=None,
        subaccount=None,
        turn_off_notification=None,
        redirect_url=None,
        enable_xumm_payment=None,
        enable_interledger_payment=None,
    ):
        """
        This function handles the initiate chimoney API.

        example:
            chimoneys = [
                {
                    "email": "test@example",
                    "phone": "+16471112222",
                    "valueInUSD": 1,
                    "twitter": "@test"
                }
            ]

            crypto_payments = [
                {
                    "xrpl": {
                        "address": "rHb9CJAWyB4rj91VRWn96DkukG4bwdtyTh",
                        "issuer": "rHb9CJAWyB4rj91VRWn96DkukG4bwdtyTh",
                        "currency": "XRP"
                    }
                }
            ]

        Args:
            chimoneys(list, required): A list of dictionaries containing the chimoney details.
            crypto_payments(list): A list of dictionaries containing the crypto payment details.
            subaccount(str): The subaccount to be used.
            turn_off_notification(bool): None by Default.
            redirect_url(str): The URL to redirect to after payment is confirmed.
            enable_xumm_payment(bool): To generate a XUMM transaction sign link.
            enable_interledger_payment(bool): To generate an Open Payment
              payment request to pay with Interledger.

        Return:
            dict: The response from the Chi Money API.
        """

        if not isinstance(chimoneys, list):
            raise ValueError("chimoneys must be a list of dictionaries")

        if not isinstance(crypto_payments, list):
            raise ValueError("crypto_payments must be a list of dictionaries")

        payload = {"chimoneys": chimoneys, "cryptoPayment": crypto_payments}
        if subaccount:
            payload["subAccount"] = subaccount
        if turn_off_notification is not None:
            payload["turnOffNotification"] = turn_off_notification
        if redirect_url:
            payload["redirect_url"] = redirect_url
        if enable_xumm_payment:
            payload["enableXUMMPayment"] = enable_xumm_payment
        if enable_interledger_payment:
            payload["enableInterledgerPayment"] = enable_interledger_payment

        return self._handle_request(
            "POST", "/v0.2/payouts/initiate-chimoney", data=payload
        )

    def wallet(self, subaccount=None, turn_off_notifications=None, wallets=None):
        """
        The function handles the payout to Chimoney Wallet

        Args:
            subaccount(str): The subaccount to be used.
            turn_off_notification(bool): None by default
            wallets(list, required): A list of dictionaries containing wallet details

        Returns:
            dict: The response from Chimoney API
        """

        if not isinstance(wallets, list):
            return ValueError("wallets must be a list of dictionaries")

        payload = {"wallets": wallets}
        if subaccount:
            payload["subAccount"] = subaccount

        if turn_off_notifications is not None:
            payload["turnOffNotifications"] = turn_off_notifications

        return self._handle_request("POST", "/v0.2/payouts/wallet", data=payload)

File: submissions/chimoney-js/modules/Payouts.js
const Joi = require("joi");
const { ValueError, TypeError } = require("../Errors");
const {
  handleRequest,
  HTTPMETHODS,
  formatJoiErrors,
} = require("../utils/helpers");

/**
 * This function handles the Chi Money airtime API.
 * @param {Array<object>} airtimes An array of object containing the airtime details
 * @example
 *  const airtimes = [
 *  	{
 *			countryToSend: "Nigeria",
 *			phoneNumber: "+2348123456789",
 *			valueInUSD: 3
 *		}
 *	]
 * @param {string?} subAccount The subAccount of the transaction
 * @returns The response from the Chi Money API
 */
async function airtime(airtimes = [], subAccount = null) {
  if (!airtimes) throw ValueError("airtimes is required");

  if (!Array.isArray(airtimes))
    throw TypeError("airtimes must be an array of objects");

  const payload = { airtime: airtimes };

  if (subAccount) payload.subAccount = subAccount;

  return handleRequest({
    method: HTTPMETHODS.POST,
    payload,
    path: "/v0.2/payouts/airtime",
  });
}

/**
 * This function handles the bank API
 * @param {Array<object>} banks An array of objects containing the bank details
 * @example
 * const banks = [
 * 		{
 * 			countryToSend: "Nigeria",
 * 			account_bank: "044",
 * 			account_number: "0690000031",
 * 			valueInUSD: 1,
 * 			reference: "1234567890"
 * 		}
 * 	]
 * @param {string?} subAccount The subAccount of the transaction
 * @returns The response from Chi Money API
 */
async function bank(banks = [], subAccount = null) {
  if (!banks) throw new ValueError("banks is required");

  if (!Array.isArray(banks))
    throw new TypeError("banks must be an array of objects");

  const payload = { bank: banks };

  if (subAccount) payload.subAccount = subAccount;

  return handleRequest({
    method: HTTPMETHODS.POST,
    payload,
    path: "/v0.2/payouts/bank",
  });
}

/**
 * This functions handles the chimoney API
 * @param {Array<object>} chimoneys An array of objects containing the chimoney details.
 * @example
 * 	const chimoneys = [
 * 		{
 * 			valueInUSD: 1,
 * 			email: "text@example.com",
 * 			twitter: "@tester"
 * 		}
 * 	]
 * @param {string?} subAccount The subAccount of the transaction
 * @returns The response from the Chi Money API
 */
async function chimoney(chimoneys = [], subAccount = null) {
  if (!chimoneys) throw new ValueError("chimoneys is required");

  if (!Array.isArray(chimoneys))
    throw new TypeError("chimoneys must be an array of objects");

  const payload = { chimoneys };

  if (subAccount) payload.subAccount = subAccount;

  return handleRequest({
    method: HTTPMETHODS.POST,
    payload,
    path: "/v0.2/payouts/chimoney",
  });
}

/**
 * This function handles mobile money API
 * @param {Array<object>} momos An array of objects containing the mobile money details
 * @example
 * const momos = [
 * 		{
 * 			countryToSend: "Nigeria",
 * 			phoneNumber: "+2348123456789",
 * 			valueInUSD: 1,
 * 			reference: "1234567890"
 * 		}
 * ]
 * @param {string?} subAccount The subAccount of the transaction
 * @returns The response from the Chi Money API
 */
async function mobileMoney(momos = [], subAccount = null) {
  if (!momos) throw new ValueError("momos is required");

  if (!Array.isArray(momos))
    throw new TypeError("momos must be an array of objects");

  const payload = { momo: momos };

  if (subAccount) payload.subAccount = subAccount;

  return handleRequest({
    method: HTTPMETHODS.POST,
    payload,
    path: "/v0.2/payouts/mobile-money",
  });
}

/**
 * This function handles the gift card API
 * @param {Array<object>} giftCards An array of objects containing the gift card details
 * @example
 * const giftCards = [
 * 		{
 * 			email: "test@example.com",
 * 			valueInUSD: 1,
 * 			redeemData: {
 * 				productId: "3",
 * 				countryCode: "NG",
 * 				valueInLocalCurrency: 1000
 * 			}
 * 		}
 * ]
 * @param {string?} subAccount The subAccount of the transaction
 * @returns The response from the Chi Money API
 */
async function giftCard(giftCards = [], subAccount = null) {
  if (!giftCards) throw new ValueError("giftCards is required");

  if (!Array.isArray(giftCards))
    throw new TypeError("gitCards must be an array of objects");

  const payload = { gift_card: giftCards };

  if (subAccount) payload.subAccount = subAccount;

  return handleRequest({
    method: HTTPMETHODS.POST,
    payload,
    path: "/v0.2/payouts/gift-card",
  });
}

/**
 * This function handles the status API
 * @param {string} chiRef The Chi Money reference
 * @param {string?} subAccount The subAccount of the transaction
 * @returns The response from the Chi Money API
 */
async function status(chiRef, subAccount = null) {
  if (!chiRef) throw new ValueError("chiRef is required");

  if (typeof chiRef !== "string")
    throw new TypeError("chiRef must be a string");

  const payload = { chiRef };

  if (subAccount) payload.subAccount = subAccount;

  return handleRequest({
    method: HTTPMETHODS.POST,
    path: "/v0.2/payouts/status",
    payload,
  });
}

/**
 * This function handles the initiate chimoney API
 * @param {Array<object>} chimoneys An array of objects containing the chimoney details
 * @example
 * const chimoneys = [
 * 		{
 * 			valueInUSD: 1,
 * 			email: "text@example.com",
 * 			twitter: "@tester"
 * 		}
 * 	]
 * @param {Array<object>} crypto_payments An array of objects containing the crypto payment details
 * @example
 * const crypto_payments = [
 * 		{
 * 			xrpl: {
 * 				address: "rHb9CJAWyB4rj91VRWn96DkukG4bwdtyTh",
 * 				issuer: "rHb9CJAWyB4rj91VRWn96DkukG4bwdtyTh",
 * 				currency: "XRP",
 *
 * 			}
 * 		}
 * 	]
 *
 * @param {boolean?} turnOffNotification if true, it turns off email notification
 * @param {string?} subAccount The subAccount of the transaction
 * @returns The response from the Chi Money API
 */
async function initiateChimoney(
  chimoneys = [],
  turnOffNotification = false,
  crypto_payments,
  subAccount = null
) {
  // Define validation schema
  const schema = Joi.object({
    turnOffNotification: Joi.boolean().default(false),
    crypto_payments: Joi.array().optional().default([]),
    chimoneys: Joi.array().required(),
  });

  // Validate input
  const { value, error } = schema.validate(
    { turnOffNotification, crypto_payments, chimoneys },
    { abortEarly: false }
  );

  // Handle validation errors
  if (error) throw new ValueError("Invalid input(s)", formatJoiErrors(error));

  const payload = { ...value };

  if (subAccount) payload.subAccount = subAccount;

  return handleRequest({
    method: HTTPMETHODS.POST,
    payload,
    path: "/v0.2/payouts/initiate-chimoney",
  });
}

module.exports = {
  mobileMoney,
  airtime,
  initiateChimoney,
  status,
  giftCard,
  chimoney,
  bank,
};
